src/keygen.cpp

```diff
@@ -440,4 +279,58 @@ void KeyGen::get_bsk2(BSKey_LWE& bsk, const SKey_base_LWE& sk_base, const SKey_b
     }    
 
     cout << "Bootstrapping2 generation: " << float(clock()-start)/CLOCKS_PER_SEC << endl;
-}
\ No newline at end of file
+}
+
+void enc_ngs(NGSFFTctxt& ct, int m, int l, int B, const SKey_boot& sk_boot)
+{
+    ModQPoly msg(Param::N,0L);
+    msg[0] = m; // msg = m (degree-0 polynomial)
+    enc_ngs(ct, msg, l, B, sk_boot);
+}
+
+void mult_poly_by_int(ModQPoly& a, const int b){
+    for(int i = 0; i < a.size(); i++)
+        a[i] *= b;
+}
+
+void enc_ngs(NGSFFTctxt& ct, const ModQPoly& m, int l, int B, const SKey_boot& sk_boot)
+{
+    if(ct.size() != l)
+        ct = NGSFFTctxt(l);
+
+    FFTPoly sk_boot_inv_fft(Param::N2p1); // f^-1 in FFT form
+    fftN.to_fft(sk_boot_inv_fft, sk_boot.sk_inv);
+    FFTPoly g_fft(Param::N2p1);
+    ModQPoly msg(m); // at each iteration i, msg will be equal to m * B^i
+    FFTPoly msg_fft(Param::N2p1);
+    FFTPoly tmp_ct(Param::N2p1);
+    vector<long> tmp_ct_long(Param::N);
+    vector<int> tmp_ct_int(Param::N);
+
+    int powerB = 1;
+
+    for (int i = 0; i < l; i++)
+    {
+        // sample random ternary vector
+        ModQPoly g(Param::N,0L);
+        Sampler::get_ternary_vector(g);
+        // FFT transform it
+        fftN.to_fft(g_fft, g);
+        // compute g * sk_boot^(-1)
+        tmp_ct = g_fft * sk_boot_inv_fft;
+        // compute g * sk_boot^(-1) + B^i * m
+        fftN.to_fft(msg_fft, msg); // msg = m * B^i
```

Comment:
This FFT can be done before the loop. Then in the loop, you can multiply `msg_fft` by `B` in the same way as you do here. See the code you removed above.

Reply:
Yes, that is true. I fixed it, thanks!